File: packages/core/src/core/feedforward/dynamic_solver.py
from sleipnir.optimization import Problem, ExitStatus

from . import eq
from core.sample.logged_sample import LoggedSample
from .gains import FeedforwardGains


class DynamicSolver:

    def __init__(self, ks: float, kv: float) -> None:
        self.ks = ks
        self.kv = kv
        self.problem = Problem()
        self.ka = self.problem.decision_variable()
        self.problem.subject_to(self.ka >= 0)

    def init_data(self, data: list[LoggedSample]) -> None:
        J = 0
        for sample in data:
            u_pred = eq.calculate_feedforward(self.ks, self.kv, self.ka, sample.velocity, sample.acceleration)
            u = sample.voltage
            J += pow(u_pred-u, 2)

        self.problem.minimize(J)

    def solve(self) -> FeedforwardGains | None:
        stat = self.problem.solve()
        if stat != ExitStatus.SUCCESS:
            # TODO: Log
            return None

        return FeedforwardGains(self.ks, self.kv, self.ka.value())

File: src/core/feedforward/DynamicSolver.cpp
//
// Created by adarw on 7/20/26.
//

#include "DynamicSolver.hpp"

#include "feedforward.hpp"

namespace sysid {
    DynamicSolver::DynamicSolver(const System& system, const FeedforwardGains& quasistaticGains) : system(system),
        quasistaticGains(quasistaticGains) {
        ka = problem.decision_variable();

        // TODO: Make an initial guess

        // constraints
        problem.subject_to(ka >= 0);
    }

    void DynamicSolver::putData(const std::shared_ptr<SampleVector>& log) {
        J = 0;
        for (size_t i = 0; i < log->N; i++) {
            const double velocity = log->getVelocity(system.systemType, i);
            const double acceleration = log->getAcceleration(system.systemType, i);

            double theta = 0;
            if (system.gravityType == GravityType::ARM && system.systemType == SystemType::POSITIONAL) {
                theta = log->getPosition(system.systemType, i);
            }

            auto u_pred = calculateFeedforward(quasistaticGains.ks, quasistaticGains.kv, ka, quasistaticGains.kg, velocity, acceleration, theta, system.gravityType);
            const auto& u = log->u(i);
            J += slp::pow(u_pred - u, 2);
        }

        problem.minimize(J);
    }

    OptimizationResult<OLSMetrics, FeedforwardGains> DynamicSolver::solve() {
        if (const slp::ExitStatus status = problem.solve(); status != slp::ExitStatus::SUCCESS) {
            throw std::runtime_error("Dynamic solver could not converge to a solution!");
        }
        return OptimizationResult<OLSMetrics, FeedforwardGains>{
            {
                0,
                J.value()
            }, {
              quasistaticGains.ks,
              quasistaticGains.kv,
              ka.value(),
              quasistaticGains.kg,
                system.gravityType,
            }
        };
    }
} // sysid